### ISSUE TITLE
Icons and splash screens generated from `ionic resources` ignored on Android

### ISSUE THREAD

**Issue by bilalakil** [driftyco/ionic-cli#1608]
#### Short description of the problem:

Even in a freshly generated project, icon and splash screens generated from `ionic resources` are being ignored in Android, where the default Cordova images are instead used. This problem does not exist on iOS.
#### What behaviour are you expecting?

Icons and splash screens generated by `ionic resources` (which comes up with the Ionic logos if no other icon or splash screen is provided) to be used in the Android app.

**Steps to reproduce:**

```
ionic start x
cd x
ionic platform add android
ionic resources
ionic build android
ionic run android
```

**Other information:** (e.g. stacktraces, related issues, suggestions how to fix, stackoverflow links, forum links, etc)

The `res` directory created in the root of the project (following the `ionic resources` command) contains the expected Ionic logo images (which are created by the `ionic resources` command as an example) in the many different sizes required by Android.

The `platform/android/res` directory, however, contains all of the default Cordova images, which appear to be the ones actually being used.

`platform/res/android/xml/config.xml` perfectly matches the expected `config.xml`, so it is copying that across as expected, but not the generated images.

**Which Ionic Version?** 1.7.14

**Run `ionic info` from terminal/cmd prompt:** (paste output below)

```
Your system information:

Cordova CLI: 6.4.0
Gulp version:  CLI version 3.9.1
Gulp local:  
Ionic Version: 1.3.2
Ionic CLI Version: 1.7.14
Ionic App Lib Version: 0.7.0
ios-deploy version: 1.9.0 
ios-sim version: 5.0.10 
OS: Mac OS X El Capitan
Node Version: v6.6.0
Xcode version: Xcode 8.1 Build version 8B62 
```

**Comment by bilalakil** [comment]
Note that my current workaround to get it built with the icons is to simply move them into the correct position. I do that like so:

```
mv platforms/android/res/{values,xml} res
rm -rf platforms/android/res
cp -r res platforms/android
ionic build android
```

However I suspect if I generate a new set of icons or prepare the Android platform again, these steps would be reversed.

**Comment by Preau** [comment]
I'm also having this exact problem, seems the res map isn't created in the correct location

**Comment by camexapps** [comment]
Yes, having the same issue just after updating Ionic and Cordova NPM libraries. 

Using Ionic 1.3 app

Ionic 2.1.4
Cordova 6.4.0

**Comment by workfluo** [comment]
Hi folks,
I have this pb also. I think I know the reason (don't worry it has nothing to do with Illuminatis ;-).
The command "ionic resources" updates config.xml. The code written is partly bad code.
Instead of "density" inside:\<splash src="resources/android/splash/drawable-land-ldpi-screen.png" density="land-ldpi"/\>
It should be "qualifier" as here: \<splash src="resources/android/splash/drawable-port-hdpi-screen.png" qualifier="port-hdpi"/\>
Then magically the build process creates the right folders and copy the right files for every screens.
So basically the CLI has to patched (PLZ supa Ionic Team).
But for know let's house party: https://www.youtube.com/watch?v=tGJEe-yGLos

**Comment by kodeine** [comment]
this issue started as soon i upgraded cordova to latest (6.4.0).

**Comment by jonatastrafaniuc** [comment]
The same problem here with ionic 2 RC2.

**Comment by alfonsojohan** [comment]
@workfluo thanks man, you solution solved this issue for me. was having this issue as well, my ionic info:

`Cordova CLI: 6.3.1
Ionic Framework Version: 2.0.0-rc.1
Ionic CLI Version: 2.1.4
Ionic App Lib Version: 2.1.2
Ionic App Scripts Version: 0.0.39
OS:
Node Version: v6.9.1`

**Comment by EmreErdogan** [comment]
I had the same issue with the following system information:

> Cordova CLI: 6.4.0
> Gulp version: CLI version 3.9.1
> Gulp local: Local version 3.9.1
> Ionic Framework Version: 1.3.2
> Ionic CLI Version: 2.0.0
> Ionic App Lib Version: 2.0.0
> ios-deploy version: 1.9.0 
> ios-sim version: 5.0.8 
> OS: Mac OS X El Capitan
> Node Version: v4.4.4
> Xcode version: Xcode 8.1 Build version 8B62

I managed to make it work by downgrading Cordova CLI from 6.4.0 to 6.3.1 by running the following command:

`npm install -g cordova@6.3.1`

and removing & adding android platform then re-building for android:

```
ionic platform remove android
ionic platform add android
ionic run android
```

**Comment by workfluo** [comment]
If i try: ionic build android i get this:
Note: /Users/Workfluo/ionic-server/wad/platforms/android/src/org/apache/cordova/splashscreen/SplashScreen.java uses or overrides a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
What does it means?
I've got this <plugin name="cordova-plugin-splashscreen" spec="~4.0.0"/>, the last version.

**Comment by jonatastrafaniuc** [comment]
I will try this from ghenry22 in https://github.com/driftyco/ionic/issues/9060:
"Upgrade to android platform 6.1.0 and this should be fixed, just updated my project, reran ionic resources and then build and all good."

**Comment by workfluo** [comment]
I have this:
Cordova CLI: 6.4.0
Gulp version:  CLI version 3.9.1
Gulp local:  
Ionic Framework Version: 2.0.0-rc.2
Ionic CLI Version: 2.1.4
Ionic App Lib Version: 2.1.2
Ionic App Scripts Version: 0.0.39
And i use this for config settings:
<splash src="resources/android/splash/port/ldpi/screen.png" platform="android" qualifier="ldpi"/> + <preference name="SplashScreen" value="screen"/>
BUT in APK the splash screens are named splash.
It depends probably on cordova or ionic cli version.
Whatever it's unstabil!

**Comment by workfluo** [comment]
I would highly recommend to open the .pkg file with an unzip tool in order to check IF in "res" folder you have any image for splash screens AND their name.

**Comment by workfluo** [comment]
GET IT. GOTTCHA YOU BUGGY BUG!
My solution with a config like:
Cordova CLI: 6.4.0 + Ionic Framework Version: 2.0.0-rc.2 + Ionic CLI Version: 2.1.4;
is basically: 
after the Command $ionic resource,
in config.xml for all the
\<splash src="resources/android/splash/drawable-land-ldpi-screen.png" density="land-ldpi"/\> 
REPLACE "density" by "qualifier" 
AND 
in <preference name="SplashScreen" value="screen"/>
REPLACE value="screen" by value="splash"

**Comment by pinindajin** [comment]
@camexapps , your solution to do `cordova platform add android@5.2.2` worked. NOTE: This is for a cordova backbone app - thought i'd post here since this seems to be more a cordova issue than an ionic issue.

**Comment by rhuanbarreto** [comment]
@EmreErdogan your solution solved my problem. Downgrading cordova solves all problems!

For the people reading this, just uninstall the current cordova before installing the older version.

```
npm uninstall -g cordova
npm install -g cordova@6.3.1
```

**Comment by workfluo** [comment]
Simply because there is no single answer.
For me the pb comes not from a plugin or a cordova version.
It all about playing with playing with the propriety name "density" vs. "qualifier" and the attribut value "screen" vs. "splash".
Than rebuild.

**Comment by fonini** [comment]
+1

**Comment by karser** [comment]
I was able to fix this issue by:
```
npm install -g cordova@6.3.1 android@5.2.2
```

**Comment by florentsuc** [comment]
I found my problem. I created a new ionic test project and the splashscreen is working. 
So I made a diff between the two config.xml files.

I was using `<preference name="SplashScreen" value="splash"/>` instead of  `<preference name="SplashScreen" value="screen"/>`.

Thus, I can tell this configuration is working:
- `ionic lib: 2.0.0-rc2`
- `ionic cli: 2.1.8`
- `cordova: 6.4.0`
- `cordova-android: 6.1.0`
- `cordova-plugin-splashscreen: 4.0.0`

All you have to do is:
- `ionic platform remove android` 
- install the correct configuration (cordova, ionic cli)
- `ionic platform add android@6.1.0`
- `ionic resources`
- check you don't have a bad config.xml (see my error)

A valid Android splashcreen config line should look like: `<splash src="resources/android/splash/drawable-port-ldpi-screen.png" density="port-ldpi"/>`

Hope this helps! :-)

**Comment by vintage** [comment]
@florentsuc What about the icons? Do they work properly for you? I can't manage to fix it and always get the default cordova icon.

**Comment by florentsuc** [comment]
@vintage Yes they do. I spent so much time to find my mistake. I can only advice you to create a new ionic blank project to find your problem.

**Comment by vintage** [comment]
@florentsuc I've been reading your comment multiple times and just realized I'm using `cordova-android@6.0`. Upgrade to `cordova-android@6.1` solved the issue. The icon & splash works great now. Thanks! :)

**Comment by dilhan2013** [comment]
Thanks @vintage updating to android @ 6.1 solved the issue

**Comment by felippepuhle** [comment]
Thanks @florentsuc, your tips helped me a lot.

Add `<engine name="android" spec="~6.1.0" />` to the **config.xml** file to make `ionic platform add android` always resolve to 6.1.0.

**Comment by balate** [comment]
@florentsuc Thanks!

**Comment by xgzl87** [comment]
npm ERR! fetch failed http://registry.npmjs.org/cordova-android/-/cordova-android-6.1.0.tgz
npm WARN retry will retry, error on last attempt: Error: connect ECONNREFUSED
npm ERR! fetch failed http://registry.npmjs.org/cordova-android/-/cordova-android-6.1.0.tgz
npm WARN retry will retry, error on last attempt: Error: connect ECONNREFUSED
npm ERR! fetch failed http://registry.npmjs.org/cordova-android/-/cordova-android-6.1.0.tgz
Error: Failed to fetch platform android@~6.1.0
Probably this is either a connection problem, or platform spec is incorrect.
Check your connection and platform name/version/URL.
Error: connect ECONNREFUSED

**Comment by ataraxus** [comment]
Same issue here
```
Cordova CLI: 6.4.0
Ionic Framework Version: 2.0.0-rc.3
Ionic CLI Version: 2.1.13
Ionic App Lib Version: 2.1.7
Ionic App Scripts Version: 0.0.46
ios-deploy version: Not installed
ios-sim version: Not installed
OS: Windows 7
Node Version: v6.9.1
Xcode version: Not installed
```

**Comment by blakholesun** [comment]
Yup! Upgrading to android@6.1 fixed it for me. Thanks @vintage !

**Comment by mattbryan9** [comment]
Upgrade to 6.1 worked. Many thanks to @florentsuc and @felippepuhle!

**Comment by clodo** [comment]
Gracias @florentsuc y @felippepuhle, the workarround worked for me!

**Comment by almothafar** [comment]
@florentsuc  and @felippepuhle 
Thank you, the workaround worked, but bad news I have a new issue, the splash screen shown as a blank, not image one, if I use `<preference name="SplashScreen" value="screen" />` it works.

**Comment by mg1075** [comment]
@florentsuc - Thanks for the instructions. I can now see my custom icon on my Android device. 

Hope this issue gets a good resolution: an app distributed with a default cordova icon gives a bad impression, even if distributed only for testing purposes.

**Comment by ceberS** [comment]
Thanks @florentsuc  !!

**Comment by almothafar** [comment]
@herkulano 
Try this: when you use `ionic resources` to generate icon and splash change it to:
<preference name="SplashScreen" value="splash"/>
then change it back to:
<preference name="SplashScreen" value="screen"/>

This how I solved the blank splash.

**Comment by herkulano** [comment]
@almothafar thanks. you forgot to post the code.

**Comment by almothafar** [comment]
@herkulano the code was there but not wrapped with ```.

**Comment by herkulano** [comment]
@almothafar thanks.

It was working already with @florentsuc post. My splash problem was the previous value of: `<preference name="SplashScreenDelay" value="10000"/>` was set at `0`. Changed it and it's now working correctly.

**Comment by paragw09ckp** [comment]
Thanks to @florentsuc to get past that frustrating issue!

Now, in my case, I am app saved with the correct icon. And also getting the splash screen shown with correct image. However, after the splash the screen goes all white. Nothing happens! :(

Any suggestions or pointers?

Your help is highly appreciated!

**Comment by patrickbussmann** [comment]
Thank you both!
@felippepuhle @florentsuc 

But why isn't the 6.1 engine set as default if it contains bugs like this?

**Comment by ea2809** [comment]
Only changing the android version, everything works fine.

**Comment by PterPmnta** [comment]
@florentsuc  What Node Version you have?

**Comment by PterPmnta** [comment]
@herkulano  Sorry, i have a problem with tha Node 6.9.X, say `sys is deprecated. Use util instead`. I wanna to know this hide using Node 7.0 or above.

**Comment by louisdoe** [comment]
I too have the issue.
I just did 
    
    cordova platform update android@6.1.0

because I was on android@6.0.0 and it seems it is the cause of the problem.

But then I go through a lot of errors on compilation, plugins errors like:

```
AdMobPlugin.java:37: error: package com.rjfun.cordova.ad does not exist

import com.rjfun.cordova.ad.GenericAdPlugin;
```

I tried `ionic plugin rm` and `ionic plugin add`all the plugins, but I still see these errors.

Can you help ?

@EmreErdogan solution doesn't seem to work.

**Comment by ScreamZ** [comment]
Any support for cordova 6.4 is coming ?

Or should I use @workfluo strategy ?

**Comment by vksgautam1** [comment]
for cordova 6.4 this is working one
https://github.com/apache/cordova-android/commit/854946e313c6b00e0087e9aa432fad664083c96a

**Comment by vksgautam1** [comment]
try this. for 6.4
https://github.com/apache/cordova-android/commit/854946e313c6b00e0087e9aa432fad664083c96a

**Comment by mhartington** [comment]
If you're splashscreens and icons are not getting used/copied, run

```
ionic platform rm android
ionic platform add android@latest --save
```

Android 6.0 (which cordova currently has pinned as the default version) had a bug with resources not being used.

**Comment by jamesvsshark** [comment]
@mhartington who's better than you?

**Comment by yadutechnologies** [comment]
@mhartington Upgrading to latest android worked for me.

**Comment by Beauchaj** [comment]
Assuming you don't mind being on the latest android platform available, add the following to your config.xml, when you install the android platform or run 'ionic state restore' it will pull the latest version of android automatically (so you don't have to rm, and add each time you build from your repo).

`<engine name="android" spec="latest"/>`

same level as platform, **not** inside the platform tag

**Comment by Fogetti** [comment]
This proposed solution does not work anymore.

My environment is the following:
```
Your system information:

Cordova CLI: 6.5.0
Ionic Framework Version: 3.0.1
Ionic CLI Version: 2.2.2
Ionic App Lib Version: 2.2.1
Ionic App Scripts Version: 1.3.0
ios-deploy version: 1.9.1
ios-sim version: 5.0.13
OS: macOS Sierra
Node Version: v7.4.0
Xcode version: Xcode 8.2.1 Build version 8C1002
```

My cordova android version looks like this:
```
Installed platforms:
  android 6.1.0
  browser 4.1.0
  ios 4.3.1
Available platforms:
  amazon-fireos ~3.6.3 (deprecated)
  blackberry10 ~3.8.0
  firefoxos ~3.6.3
  osx ~4.0.1
  webos ~3.7.0
```

And whenever I run `ionic resources` and `ionic build android` consecutively the following happens: the resources are properly generated in the `resources` folder but none of them copied over to the platforms folder.

And even when I copy the `resources` folder to the platforms folder manually, the cordova build does not honor my resources and simply excludes them from the APK.

Also, the config.xml under `platforms/android/res/xml` is expecting the icon and splash screen to be under `resources` like this:
```xml
<?xml version='1.0' encoding='utf-8'?>
<widget id="com.ionicframework.myapp" version="0.0.1" xmlns="http://www.w3.org/ns/widgets" xmlns:cdv="http://cordova.apache.org/ns/1.0">
    <feature name="AppVersion">
        <param name="android-package" value="uk.co.whiteoctober.cordova.AppVersion" />
    </feature>
    <feature name="Device">
        <param name="android-package" value="org.apache.cordova.device.Device" />
    </feature>
    <feature name="InAppBrowser">
        <param name="android-package" value="org.apache.cordova.inappbrowser.InAppBrowser" />
    </feature>
    <feature name="SplashScreen">
        <param name="android-package" value="org.apache.cordova.splashscreen.SplashScreen" />
        <param name="onload" value="true" />
    </feature>
    <feature name="StatusBar">
        <param name="android-package" value="org.apache.cordova.statusbar.StatusBar" />
        <param name="onload" value="true" />
    </feature>
    <feature name="Keyboard">
        <param name="android-package" onload="true" value="org.apache.cordova.labs.keyboard.Keyboard" />
    </feature>
    <name>Tado Installer App</name>
    <description>An awesome Ionic/Cordova app.</description>
    <author email="hi@ionicframework" href="http://ionicframework.com/">Ionic Framework Team</author>
    <content src="index.html" />
    <access origin="*" />
    <allow-navigation href="http://ionic.local/*" />
    <allow-intent href="http://*/*" />
    <allow-intent href="https://*/*" />
    <allow-intent href="tel:*" />
    <allow-intent href="sms:*" />
    <allow-intent href="mailto:*" />
    <allow-intent href="geo:*" />
    <icon src="resources/android/icon/drawable-xhdpi-icon.png" />
    <allow-intent href="market:*" />
    <icon density="ldpi" src="resources/android/icon/drawable-ldpi-icon.png" />
    <icon density="mdpi" src="resources/android/icon/drawable-mdpi-icon.png" />
    <icon density="hdpi" src="resources/android/icon/drawable-hdpi-icon.png" />
    <icon density="xhdpi" src="resources/android/icon/drawable-xhdpi-icon.png" />
    <icon density="xxhdpi" src="resources/android/icon/drawable-xxhdpi-icon.png" />
    <icon density="xxxhdpi" src="resources/android/icon/drawable-xxxhdpi-icon.png" />
    <splash density="land-ldpi" src="resources/android/splash/drawable-land-ldpi-screen.png" />
    <splash density="land-mdpi" src="resources/android/splash/drawable-land-mdpi-screen.png" />
    <splash density="land-hdpi" src="resources/android/splash/drawable-land-hdpi-screen.png" />
    <splash density="land-xhdpi" src="resources/android/splash/drawable-land-xhdpi-screen.png" />
    <splash density="land-xxhdpi" src="resources/android/splash/drawable-land-xxhdpi-screen.png" />
    <splash density="land-xxxhdpi" src="resources/android/splash/drawable-land-xxxhdpi-screen.png" />
    <splash density="port-ldpi" src="resources/android/splash/drawable-port-ldpi-screen.png" />
    <splash density="port-mdpi" src="resources/android/splash/drawable-port-mdpi-screen.png" />
    <splash density="port-hdpi" src="resources/android/splash/drawable-port-hdpi-screen.png" />
    <splash density="port-xhdpi" src="resources/android/splash/drawable-port-xhdpi-screen.png" />
    <splash density="port-xxhdpi" src="resources/android/splash/drawable-port-xxhdpi-screen.png" />
    <splash density="port-xxxhdpi" src="resources/android/splash/drawable-port-xxxhdpi-screen.png" />
    <preference name="loglevel" value="DEBUG" />
    <preference name="webviewbounce" value="false" />
    <preference name="UIWebViewBounce" value="false" />
    <preference name="DisallowOverscroll" value="true" />
    <preference name="android-minSdkVersion" value="16" />
    <preference name="BackupWebStorage" value="none" />
    <preference name="SplashMaintainAspectRatio" value="true" />
    <preference name="FadeSplashScreenDuration" value="300" />
    <preference name="SplashShowOnlyFirstTime" value="false" />
    <preference name="SplashScreen" value="screen" />
    <preference name="SplashScreenDelay" value="3000" />
</widget>
```

Which is of course wrong! Because there is no such folder under `platforms`.

At this point I have no clue who does what and who to blame, because not clear: why isn't the `ionic resources` command copy over the resources, and secondly why are the resources excluded from the APK, and thirdly, why are the config xml entries referencing the wrong locations?

I see that there is an open issue in ionic which looks painfully similar to what I described: https://github.com/driftyco/ionic-cli/issues/1857

So I just leave this here for people who later stumble upon this problem.

**Comment by frey1esm** [comment]
I got `Error: cordovaProject.projectConfig.getFileResources` after running

```
ionic platform rm android
ionic platform add android@latest --save
```

**Comment by entymon** [comment]
Thanks a lot @florentsuc. I updated android platform to 6.1.3 but seems like only 6.1.0 works well.
